File: algorithms/code/leetcode/lc046_permutations/LC046Permutations.java
package leetcode.lc046_permutations;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;

/**
 * https://leetcode.com/problems/permutations/
 * #medium
 */
public final class LC046Permutations {
    public void permute(final List<Integer> prefix, final Set<Integer> remaining,
            final List<List<Integer>> permutations) {
        if (remaining.size() == 0) {
            permutations.add(new ArrayList<>(prefix));
        } else {
            int size = prefix.size();
            for (int value : remaining) {
                prefix.add(value);
                Set<Integer> reduced = new HashSet<>(remaining);
                reduced.remove(value);
                permute(prefix, reduced, permutations);
                prefix.remove(size);
            }
        }
    }

    public List<List<Integer>> permute(final int[] nums) {
        List<List<Integer>> permutations = new LinkedList<>();
        Set<Integer> remaining = new HashSet<>();
        for (int value : nums) {
            remaining.add(value);
        }
        permute(new LinkedList<>(), remaining, permutations);
        return permutations;
    }
}
package leetcode.lc046_permutations;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.io.Serializable;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public final class LC046PermutationsTests {
    private static class IntegerListComparator implements Comparator<List<Integer>>, Serializable {
        @Override
        public int compare(final List<Integer> l1, final List<Integer> l2) {
            if (l1.size() < l2.size()) {
                return -1;
            }
            if (l1.size() > l2.size()) {
                return 1;
            }
            for (int i = 0; i < l1.size(); i++) {
                if (l1.get(i) < l2.get(i)) {
                    return -1;
                }
                if (l1.get(i) > l2.get(i)) {
                    return 1;
                }
            }
            return 0;
        }
    }

    @Test
    public void testExample() throws Exception {
        int[] nums = { 1, 2, 3 };
        int[][] expected = { { 1, 2, 3 }, { 1, 3, 2 }, { 2, 1, 3 }, { 2, 3, 1 }, { 3, 1, 2 }, { 3, 2, 1 } };
        List<List<Integer>> result = new LC046Permutations().permute(nums);
        Collections.sort(result, new IntegerListComparator());
        assertEquals(expected.length, result.size());
        for (int i = 0; i < expected.length; i++) {
            assertEquals(expected[i].length, result.get(i).size());
            for (int j = 0; j < expected[i].length; j++) {
                assertEquals(expected[i][j], result.get(i).get(j).intValue());
            }
        }
    }
}

File: algorithms/code/leetcode/lc046_permutations/lc046_permutations.rs
// https://leetcode.com/problems/permutations/
// #medium

use std::collections::HashSet;

pub struct Solution;

impl Solution {
    fn permute_internal(
        mut prefix: Vec<i32>,
        remaining: HashSet<i32>,
        permutations: &mut Vec<Vec<i32>>,
    ) {
        if remaining.is_empty() {
            permutations.push(prefix.clone());
        } else {
            for value in remaining.iter() {
                prefix.push(*value);
                let mut reduced = remaining.clone();
                reduced.remove(value);
                Self::permute_internal(prefix.clone(), reduced, permutations);
                prefix.pop();
            }
        }
    }

    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut permutations = vec![];
        let mut remaining = HashSet::new();
        for value in nums.iter() {
            remaining.insert(*value);
        }
        Self::permute_internal(Vec::new(), remaining, &mut permutations);
        permutations
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use std::cmp::Ordering;

    pub fn compare_integer_vectors(l1: &Vec<i32>, l2: &Vec<i32>) -> Ordering {
        if l1.len() < l2.len() {
            return Ordering::Less;
        }
        if l1.len() > l2.len() {
            return Ordering::Greater;
        }
        for (a, b) in l1.iter().zip(l2.iter()) {
            if a < b {
                return Ordering::Less;
            }
            if a > b {
                return Ordering::Greater;
            }
        }
        Ordering::Equal
    }

    #[test]
    fn test_example() {
        let nums = vec![1, 2, 3];
        let expected = vec![
            vec![1, 2, 3],
            vec![1, 3, 2],
            vec![2, 1, 3],
            vec![2, 3, 1],
            vec![3, 1, 2],
            vec![3, 2, 1],
        ];
        let mut result = Solution::permute(nums);
        result.sort_by(compare_integer_vectors);
        assert_eq!(expected, result);
    }

    #[test]
    fn test_comparator() {
        assert_eq!(
            compare_integer_vectors(&vec![1, 2], &vec![1, 2, 3]),
            Ordering::Less
        );
        assert_eq!(
            compare_integer_vectors(&vec![1, 2, 3], &vec![1, 2]),
            Ordering::Greater
        );
        assert_eq!(
            compare_integer_vectors(&vec![1, 2], &vec![1, 2]),
            Ordering::Equal
        );
        assert_eq!(
            compare_integer_vectors(&vec![0, 2], &vec![1, 2]),
            Ordering::Less
        );
        assert_eq!(
            compare_integer_vectors(&vec![2, 3], &vec![1, 2]),
            Ordering::Greater
        );
    }
}
